server/repository/src/db_diesel/log_row.rs

```diff
@@ -0,0 +1,90 @@
+use super::{log_row::log::dsl as log_dsl, StorageConnection};
+
+use crate::{
+    db_diesel::{invoice_row::invoice, store_row::store},
+    repository_error::RepositoryError,
+    user_account,
+};
+
+use chrono::NaiveDateTime;
+use diesel::prelude::*;
+use diesel_derive_enum::DbEnum;
+
+table! {
+    log (id) {
+        id -> Text,
+        log_type -> crate::db_diesel::log_row::LogTypeMapping,
+        user_id -> Nullable<Text>,
+        store_id -> Nullable<Text>,
+        record_id -> Nullable<Text>,
+        datetime -> Timestamp,
+    }
+}
+
+joinable!(log -> user_account (user_id));
+joinable!(log -> store (store_id));
+joinable!(log -> invoice (record_id));
+
+#[derive(DbEnum, Debug, Clone, PartialEq, Eq)]
+#[DbValueStyle = "SCREAMING_SNAKE_CASE"]
+pub enum LogType {
+    UserLoggedIn,
+    InvoiceCreated,
+    InvoiceStatusAllocated,
+    InvoiceStatusPicked,
+    InvoiceStatusShipped,
+    InvoiceStatusDelivered,
+    InvoiceStatusVerified,
+}
+
+#[derive(Clone, Queryable, Insertable, AsChangeset, Debug, PartialEq)]
+#[changeset_options(treat_none_as_null = "true")]
+#[table_name = "log"]
+pub struct LogRow {
+    pub id: String,
+    pub log_type: LogType,
+    pub user_id: Option<String>,
+    pub store_id: Option<String>,
+    pub record_id: Option<String>,
+    pub datetime: NaiveDateTime,
+}
+
+pub struct LogRowRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> LogRowRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        LogRowRepository { connection }
+    }
+
+    #[cfg(feature = "postgres")]
+    pub fn upsert_one(&self, row: &LogRow) -> Result<(), RepositoryError> {
```

Comment:
Will we update log records? Should we just do insert instead?

Reply:
BTW: With insert you don't need different code for postgres & sqlite.
`insert_into` works for both, you just can't use `on_conflict` with sqlite